File: C++/1254. Number of Closed Islands.cpp
class Solution {
public:
    bool dfs(vector<vector<int>>& grid, int x, int y) {
        if (x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size()) return false;
        if (grid[x][y] == 1) return true;

        grid[x][y] = 1;

        bool a = dfs(grid, x + 1, y), b = dfs(grid, x - 1, y);
        bool c = dfs(grid, x, y + 1), d = dfs(grid, x, y - 1);

        return a && b&& c && d;
    }

    int closedIsland(vector<vector<int>>& grid) {
        int count = 0;

        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 0 && this->dfs(grid, i, j)) count++;
            }
        }

        return count;
    }
};

File: Python/1254. Number of Closed Islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        h, w = len(grid), len(grid[0])
        count = 0

        def dfs(x: int, y: int) -> bool:
            if x < 0 or x >= h or y < 0 or y >= w: return False
            if grid[x][y] == 1: return True

            grid[x][y] = 1

            a, b = dfs(x, y + 1), dfs(x, y - 1)
            c, d = dfs(x + 1, y), dfs(x - 1, y)

            return a and b and c and d
        
        for i in range(h):
            for j in range(w):
                if grid[i][j] == 0 and dfs(i, j): count += 1

        return count
